File: python/quickSort.py
import random

class QuickSort:
    def __init__(self,ip_arr):
        self.ip_arr = ip_arr
    
    def __shuffle(self):
        for i in range(1,len(self.ip_arr)):
            j = random.randint(0,i)
            self.__swap(i,j)
    
    def __swap(self,i,j):
        temp = self.ip_arr[i]
        self.ip_arr[i] = self.ip_arr[j]
        self.ip_arr[j] = temp

    def __isSorted(self):
        for i in range(1,len(self.ip_arr)):
            if self.ip_arr[i] < self.ip_arr[i-1]:
                return False
        return True

    def __partition(self,lo,hi):
        pivot = lo
        i = pivot+1
        j = hi
        while(True):
            while(self.ip_arr[pivot] > self.ip_arr[i]):
                if i == hi:
                    break
                i+=1
            while(self.ip_arr[pivot] < self.ip_arr[j]):
                if j == lo:
                    break
                j-=1
            if(i>=j):
                self.__swap(pivot,j)
                pivot = j
                return pivot
            self.__swap(i,j)

    def __sort(self,lo,hi):
        if hi<=lo:
            return
        pivot = self.__partition(lo,hi)
        self.__sort(lo,pivot-1)
        self.__sort(pivot+1,hi)

    def sort(self):
        self.__shuffle()
        self.__sort(0,len(self.ip_arr)-1)
        if self.__isSorted():
            print(self.ip_arr)
        else:
            print("array unsorted")
            print(self.ip_arr)

if __name__ == "__main__":
    ip_arr = [31, 18, 20, 17, 18, 36, 44, 35, 22, 14]
    quickSort = QuickSort(ip_arr)
    quickSort.sort()

File: java/src/com/company/QuickSort.java
package com.company;

import java.util.Random;

public class QuickSort {
    // java use tuned quicksort for primitive types. Since the performance will be the
    // first thing to consider when sorting primitive types.

    private static void shuffle(Comparable[] a){
        // need shuffle the array first to get better performance
        Random rand = new Random();
        for (int i = 1; i < a.length; i++){
            int j = rand.nextInt(i);
            swap(a,i,j);
        }
        for(int i = 0; i < a.length; i++){
            System.out.println(a[i]);
        }
        System.out.println("Shuffled array shown above.");

    }

    private static int partition(Comparable[] a,int lo, int hi){
        // find the pivot that no smaller on its right and no larger entry on its left
        int pivot = lo;
        int i = pivot;
        int j = hi+1;
        while (true) {
            while(less(a[++i],a[pivot])){
                if(i==hi){
                    break;
                }
            }
            while(less(a[pivot],a[--j])){
                if(j==lo){
                    break;
                }
            }
            if(i>=j){
                swap(a,pivot,j);
                pivot = j;
                return pivot;
            }
            swap(a,j,i);
        }

    }

    private static void sort(Comparable[] a,int lo,int hi){
        if (hi <= lo){return;}

        int j = partition(a,lo,hi);
        sort(a,lo,j-1);
        sort(a,j+1,hi);

    }

    private static void swap(Comparable[] a,int i,int j){
        Comparable temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    private static boolean less(Comparable w, Comparable v){
        return w.compareTo(v)<0;
    }

    private static boolean isSort(Comparable[] a){
        for (int i = 1; i < a.length; i++){
            if(less(a[i],a[i-1])){
                return false;
            }
        }
        return true;
    }

    private static void show(Comparable[] a){
        for(int i = 0; i < a.length; i++){
            System.out.println(a[i]);
        }
    }

    public static void sort(Comparable[] a){
        shuffle(a);
        int size = a.length;
        sort(a,0,size-1);
        if(isSort(a)){
            show(a);
        }else{
            System.out.println("array unsorted.");
        }
    }



    public static void main(String[] args) {
        Integer[] input = new Integer[]{31, 18, 20, 17, 18, 36, 44, 35, 22, 14,19090,19090};
        sort(input);
    }

}
